openshift/machine-config-operator
# OCPBUGS-45496: Prevent undesired MOSBs from building

**- What I did**

reconciler: only build mosb's currently targeted by the MCP

In OCPBUGS-45496, here are the steps which lead to the failure of the creation of an MOSB upon the addition of a new MC:

1) An MC with erroneous contents not caught by the API validations creates a rendered-MC which triggers a MOSB build
and fails as expected with an error
2) This erroneous MOSB keeps getting added to the rate limited worker
   queue till it hits the max retries. Then it is forgotten from the
queue and is subjected to a backoff time to get added back again 3) In the meantime if the erroneous MC is deleted and a new valid MC is
   added targetting the same MCP, a valid MOSB build starts
4) When the erroneous mosb enters the queue again and sees that there
   already exists another not successfull MOSB in build, it cancels all
other builds.

Hence the valid MOSB is cancelled and the erroneous MOSB is re-triggerred again. Since it will never be able to start the build and fail again the steps 1 and 2 keep happening.
Any new MC will fail to successfully create a MOSB and trigger a build.

Thus the solution is to check upon the MOSB sync whether the MCP and rendered-MC that the MOSB targets even exists anymore. If not we don't need to build it.

**- How to verify it**

1) Create an erroneous MC which is still valid by the API server. Example (adding an unsupported extension as a systemd unit)
2) Wait till a rendered-MC and MOSB appears and fails
3) Remove the erroneous MC
4) Add a valid MC (for example supported extension or a core user ssh keys change)
5) The valid MC should trigger a MOSB and build successfully

**- Description for the changelog**
